### ISSUE TITLE
perf: disabling split queue gives performance improvement (cpu ~1.2%)

### ISSUE THREAD

**Issue by ajstorm** [cockroachdb/cockroach#136366]
When running sysbench `oltp_read_write` we're seeing a 1.2% improvement when disabling the split queue using:

```
set cluster setting kv.split_queue.enabled = false;
```

Run info can be found [here](https://console.cloud.google.com/storage/browser/perf-efficiency-team/sysbench-oltp_read_write-uniform.10x10mrows.64threads.600sec.1500wh.10mwh.ajstorm-kv.split_queue.enabled;tab=objects?project=cockroach-ephemeral&pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&prefix=&forceOnObjectsSortingFiltering=false).

Jira issue: CRDB-45000

Epic CRDB-55523

**Comment by tbg** [comment]
Noting that this is on a three-node setup. @kvoli if we had anywhere for the split replicas to rebalance to it would be hard to argue with the split (after all, it could help balance CPU across more nodes) but in a three node setup, is there any benefit to performing the split?

**Comment by kvoli** [comment]
Only benefit I could think of w.r.t distribution is there needs to be at least node count splits (for user keyspace), for a read heavy workload, otherwise the leaseholder node might bottleneck the workload on CPU.

I could see at speculative argument about per-range throughput having some upper limit, where splitting would be helpful. Unsure if that is really true, although I recall it being brought up before.

Was the throughput higher with the split queue disabled as well?

**Comment by ajstorm** [comment]
> Was the throughput higher with the split queue disabled as well?

Yes, qps was higher by approximately 1.2%.  The labeling of these issues is a bit confusing.  Since we're cpu bound, we're directly mapping the throughput improvement to a CPU reduction, which isn't accurate, but it's a reasonable approximation for now.

**Comment by tbg** [comment]
Trying things out with the prototype https://github.com/cockroachdb/cockroach/pull/140596 which splits based on approximated disk usage rather than logical bytes.

This gave the expected results, i.e. splits reasonably correspond to on-disk size:	

```
SET CLUSTER SETTING kv.range_split.by_load.enabled = false;
```

```
roachprod@localhost:26257/defaultdb> select count(1) from [ show ranges from database sysbench ];
  count
---------
     21
(1 row)
```

```
roachprod@localhost:26257/defaultdb> set cluster setting kv.split_queue.enabled = true;
SET CLUSTER SETTING

Time: 3ms total (execution 3ms / network 0ms)
```

(wait a few minutes for the number below to stabilize)

```
roachprod@localhost:26257/defaultdb> select count(1) from [ show ranges from database sysbench ];
  count
---------
     87
(1 row)
```

```
$ roachprod ssh tobias-split:1-3 -- df -h /mnt/data1
tobias-split:[1 2 3]: df -h /mnt/data1 3/3
   1: 	<ok>
	Filesystem      Size  Used Avail Use% Mounted on
	/dev/nvme0n1    369G   22G  347G   6% /mnt/data1

   2: 	<ok>
	Filesystem      Size  Used Avail Use% Mounted on
	/dev/nvme0n1    369G   22G  347G   6% /mnt/data1

   3: 	<ok>
	Filesystem      Size  Used Avail Use% Mounted on
	/dev/nvme0n1    369G   22G  347G   6% /mnt/data1
```

So here, after size-based splitting is done, each range accounts for ~253MB (22GB/87).

I feel like I'm missing something though. Each store reports ~22G of live bytes, which are logical bytes:

<img width="959" alt="image" src="https://github.com/user-attachments/assets/11400665-7596-4091-bfd4-94b5181de750" />

I thought the point was that this data compresses very well. So shouldn't there be **more** LiveBytes relative to what's on each disk? The data is definitely snappy compressed:

<img width="963" alt="image" src="https://github.com/user-attachments/assets/352a4069-2b6b-4e37-85e2-8818746dbef1" />

I am hitting EOD, but I have some question marks about how this matches up with the initial experiment.

Actually, re-reading this, there's a basic math error: I am saying that each disk has <50GB in it. At 188 ranges, that's ~265MB per range. So the initial premise that we were oversplitting because of compression does not hold.

Instead, just splitting the ranges down to size is working to our detriment: for this workload to perform optimally, we'd really want everything in just a few ranges, as is the case when the split queue is off.

**Comment by ajstorm** [comment]
I'm wondering if this whole time we were chasing an artifact of inter-run variability.  Now that we've gotten things much more stable, I can no longer reproduce the performance difference:

No splits:

```
Mean: 21738.72
Median: 21746.97
Standard Deviation: 92.47
Coefficient of variance: .0042
```

Splits:

```
Mean: 21700.75
Median: 21679.48
Standard Deviation: 179.37
Coefficient of variance: .0082
```

The difference is now 0.3%, which is well within the inter-run variability.  @tbg I'm thinking that we may want to pause any further investigation on this one for now (and maybe forever).

**Comment by tbg** [comment]
Some Grafana poking on this.

# two back-to-back runs with splits

10:30 and 11:00 should be in the main oltp_read_write segment.

The version with splits balances load earlier in the run (makes sense - more ranges to move around):

<img width="935" alt="Image" src="https://github.com/user-attachments/assets/73ed7a1e-1880-47d7-a4d5-1b30195b4385" />

Thanks to splitting, we have 146 replicas per node. I've checked the timing on both of the displayed runs, and the splits finished well ahead of the "real" workload starting. For example, in the first run, the replica count stabilizes at 10:22, but the workload starts only around 10:25.

<img width="936" alt="Image" src="https://github.com/user-attachments/assets/383af2b9-4432-4c95-a1fa-3763333b3ea3" />

during the benchmark, the raft scheduler latency p99 is around 2ms:

<img width="462" alt="Image" src="https://github.com/user-attachments/assets/da06da7e-d868-4d3a-bb47-eb7287b9a3e5" />

# three back-to-back runs without splits

CPU balance is not as well, note in particular the middle run, though most of this was likely during the three-minute warm-up interval (and so didn't count in the result).

<img width="943" alt="Image" src="https://github.com/user-attachments/assets/a02971ee-89d2-4476-b28b-7785901f3baa" />

We have only ~75 replicas per node (in reality, I think we "really" have two replicas per sysbench table, so a total of 20, though I did not verify this in these runs)

<img width="935" alt="Image" src="https://github.com/user-attachments/assets/6ddddcd0-fbfd-418e-8db9-33ea9631c7bc" />

The raft scheduler latency is a little lower - a fairly consistent ~1.68ms:

<img width="464" alt="Image" src="https://github.com/user-attachments/assets/84f75149-a1d1-49ba-b235-8c94a579a3c3" />

# kv exec latency direct comparison

splits: p90 ~ 4.1ms, p99 ~ 10.4ms, p99.9 ~ 21.6ms
unsplit: p90 ~ 3.4ms, p99 ~ 10ms, p99.9 ~ 21.7ms

Not so different in the outliers, but significantly better in the p90.

<img width="931" alt="Image" src="https://github.com/user-attachments/assets/0db25fe7-9038-4d58-9935-822f36faf392" />

<img width="909" alt="Image" src="https://github.com/user-attachments/assets/a7caa9ed-a85a-4f91-aed1-90ac59771881" />

We also see this in these sample printouts.

```
# split
Latency (ms):
         min:                                   12.98
         avg:                                   54.09
         max:                                  638.30
         95th percentile:                       84.47
         sum:                             38399835.23

# no split
Latency (ms):
         min:                                   11.49
         avg:                                   47.38
         max:                                  181.17
         95th percentile:                       75.82
         sum:                             38400008.92
```

The large difference in pmax stands out, I'll need to look into that more. I wonder if it is a joint feature of the split case, and could suggest some interference with splits after all.

Still, so far my going theory is that we would benefit from more batching (fewer ranges) in this workload.

**Comment by tbg** [comment]
Another piece of soft evidence: buffered writes gives a much larger perf improvement (in absolute qps terms) on sysbench-settings than on sysbench. One difference is that sysbench-settings has much larger ranges. Buffered writes results in larger batches hitting the fewer replicas together, so that they can be batched in raft ready handling. This is just conjecture at this point but shows that it will be well worth it to investigate further.
